File: src/main/java/com/company/apimgmt/repository/SystemRepository.java
package com.company.apimgmt.repository;

import com.company.apimgmt.entity.SystemEntity;
import io.smallrye.mutiny.Uni;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import io.vertx.mutiny.pgclient.PgPool;
import io.vertx.mutiny.sqlclient.Row;
import io.vertx.mutiny.sqlclient.RowSet;
import io.vertx.mutiny.sqlclient.Tuple;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@ApplicationScoped
public class SystemRepository {
    
    @Inject
    PgPool client;
    
    public Uni<SystemEntity> create(SystemEntity system) {
        return client.preparedQuery(
            "INSERT INTO systems (id, name, description, created_at, updated_at) " +
            "VALUES ($1, $2, $3, $4, $5) RETURNING *")
            .execute(Tuple.of(
                system.id,
                system.name,
                system.description,
                system.createdAt,
                system.updatedAt
            ))
            .onItem().transform(rows -> mapToSystemEntity(rows.iterator().next()));
    }
    
    public Uni<SystemEntity> findById(UUID id) {
        return client.preparedQuery("SELECT * FROM systems WHERE id = $1")
            .execute(Tuple.of(id))
            .onItem().transform(rows -> {
                var iterator = rows.iterator();
                return iterator.hasNext() ? mapToSystemEntity(iterator.next()) : null;
            });
    }
    
    public Uni<SystemEntity> findByName(String name) {
        return client.preparedQuery("SELECT * FROM systems WHERE name = $1")
            .execute(Tuple.of(name))
            .onItem().transform(rows -> {
                var iterator = rows.iterator();
                return iterator.hasNext() ? mapToSystemEntity(iterator.next()) : null;
            });
    }
    
    public Uni<List<SystemEntity>> findAll() {
        return client.query("SELECT * FROM systems ORDER BY name")
            .execute()
            .onItem().transform(this::mapToSystemEntityList);
    }
    
    public Uni<SystemEntity> update(SystemEntity system) {
        return client.preparedQuery(
            "UPDATE systems SET name = $1, description = $2, updated_at = $3 " +
            "WHERE id = $4 RETURNING *")
            .execute(Tuple.of(
                system.name,
                system.description,
                system.updatedAt,
                system.id
            ))
            .onItem().transform(rows -> {
                var iterator = rows.iterator();
                return iterator.hasNext() ? mapToSystemEntity(iterator.next()) : null;
            });
    }
    
    public Uni<Boolean> delete(UUID id) {
        return client.preparedQuery("DELETE FROM systems WHERE id = $1")
            .execute(Tuple.of(id))
            .onItem().transform(rows -> rows.rowCount() > 0);
    }
    
    public Uni<Long> countApisBySystemId(UUID systemId) {
        return client.preparedQuery("SELECT COUNT(*) FROM apis WHERE system_id = $1")
            .execute(Tuple.of(systemId))
            .onItem().transform(rows -> {
                var iterator = rows.iterator();
                return iterator.hasNext() ? iterator.next().getLong(0) : 0L;
            });
    }
    
    private SystemEntity mapToSystemEntity(Row row) {
        SystemEntity entity = new SystemEntity();
        entity.id = row.getUUID("id");
        entity.name = row.getString("name");
        entity.description = row.getString("description");
        entity.createdAt = row.getLocalDateTime("created_at");
        entity.updatedAt = row.getLocalDateTime("updated_at");
        return entity;
    }
    
    private List<SystemEntity> mapToSystemEntityList(RowSet<Row> rows) {
        List<SystemEntity> list = new ArrayList<>();
        for (Row row : rows) {
            list.add(mapToSystemEntity(row));
        }
        return list;
    }
}

File: apimgmt/repositories/system_repository.py
from sqlalchemy.orm import Session
from sqlalchemy import func
import uuid
from typing import List, Optional, Union

from apimgmt.models.system import System


class SystemRepository:
    def __init__(self, db: Session):
        self.db = db
    
    def create(self, system: System) -> System:
        """创建系统"""
        # 系统ID会在模型的默认值中设置，不需要在这里设置
        self.db.add(system)
        self.db.commit()
        self.db.refresh(system)
        return system
    
    def find_by_id(self, system_id: Union[uuid.UUID, str]) -> Optional[System]:
        """根据ID查找系统"""
        return self.db.query(System).filter(System.id == system_id).first()
    
    def find_by_name(self, name: str) -> Optional[System]:
        """根据名称查找系统"""
        return self.db.query(System).filter(System.name == name).first()
    
    def find_all(self) -> List[System]:
        """查找所有系统"""
        return self.db.query(System).order_by(System.name).all()
    
    def update(self, system: System) -> Optional[System]:
        """更新系统"""
        existing_system = self.find_by_id(system.id)
        if existing_system:
            for key, value in system.__dict__.items():
                if key != '_sa_instance_state':
                    setattr(existing_system, key, value)
            self.db.commit()
            self.db.refresh(existing_system)
            return existing_system
        return None
    
    def delete(self, system_id: Union[uuid.UUID, str]) -> bool:
        """删除系统"""
        system = self.find_by_id(system_id)
        if system:
            self.db.delete(system)
            self.db.commit()
            return True
        return False
    
    def count_apis_by_system_id(self, system_id: Union[uuid.UUID, str]) -> int:
        """统计系统的API数量"""
        from apimgmt.models.api import Api
        return self.db.query(func.count(Api.id)).filter(Api.system_id == system_id).scalar() or 0
